### PR TITLE
tgui: Typescript and Jest update

### PR DESCRIPTION
**Pull request by stylemistake**
## TypeScript

Adds proper typescript support to tgui.

You don't have to do anything extra to start benefitting from strong static types in normal javascript.

If you need access to all of the TypeScript in your UI, rename the file you're working on to `.tsx` (or `.ts` if you're not writing react/inferno components), and you will have access to the whole set of TypeScript features.

- [x] Add typescript
- ~~Type check normal JS files~~ (not feasible)
- [x] Add strong typing to some of tgui components
- [x] Add a way to declare types for backend data
- [x] Add strong typing to some of the modules in the `common` package
- ~~Launch `tgui-dev-server` via `ts-node`~~ (will be revisited later)

### Why is TypeScript good?

Because you, c*ders, will be reprimanded for type violation and bad code automatically and more efficiently.

## ESLint

Fixed ESLint integration with VSCode (partially because now we have a Yarn PnP compatible typescript compiler, which eslint needed).

VSCode will now highlight all problematic pieces of code automatically, and you can auto-fix many of these problems:

- Right click on code
- Source actions...
- Fix all

## Jest

You can now write and run unit tests in tgui.

It's quite simple: create a file ending in `.test.ts` or `.spec.ts` (usually with the same filename as the file you're testing), and create a test case:

```js
test('something', () => {
  expect('a').toBe('a');
});
```

To run the tests, type the following into the terminal:

```
bin/tgui --test
```

There is an example test in `packages/common/react.spec.ts`.

You can read more about Jest here: https://jestjs.io/docs/en/getting-started

Note, that there is still no real solution to test UIs for now, even though a lot of the support is here (jest + jsdom). That's going to come later. You can however test simple functions, and are encouraged to do so!

## Obligatory stale meme

<details>
<summary>Don't click</summary>

image

</details>

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
absolutely based

**Comment by Mothblocks** [comment]
YOOOOOOOOOOOOO

**Comment by actioninja** [comment]
https://www.youtube.com/watch?v=qDD-iYkHBhc

File: tgui/packages/common/react.ts [hunk 1]
```diff
@@ -39,7 +36,7 @@ export const normalizeChildren = children => {
  * Shallowly checks if two objects are different.
  * Credit: https://github.com/developit/preact-compat
  */
-export const shallowDiffers = (a, b) => {
+export const shallowDiffers = (a: any, b: any) => {
```
Comment (Mothblocks) [inline tgui/packages/common/react.ts:39]:
```suggestion
export const shallowDiffers = (a: object, b: object) => {
```

File: tgui/packages/common/react.ts [hunk 2]
```diff
@@ -66,8 +63,14 @@ export const pureComponentHooks = {
...
   return value !== undefined
     && value !== null
     && typeof value !== 'boolean';
 };
+
+/**
+ * A common case in tgui, when you pass a value conditionally, these are
+ * the types that can fall through the condition.
+ */
+export type BooleanLike = number | boolean | null | undefined;
```
Comment (Mothblocks) [inline tgui/packages/common/react.ts:76]: Why is this necessary? Why can't we just pass true/false and cut the truthiness?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/common/react.ts (hunk 2 above)
Comment (stylemistake) [reply tgui/packages/common/react.ts:76]: Byond sends 0/1 from the backend, so this is necessary to cut down on `Boolean()` or `!!` casts

Comment (Mothblocks) [reply tgui/packages/common/react.ts:76]: Ah, fair enough.

Might be a good case for enums, though?

Comment (stylemistake) [reply tgui/packages/common/react.ts:76]: Could you povide an example?

Comment (Mothblocks) [reply tgui/packages/common/react.ts:76]:
```ts
enum ByondBool {
    True = 1,
    False = 0,
}
```

Comment (stylemistake) [reply tgui/packages/common/react.ts:76]: I see what you mean, but we should implement this later, I don't want to delay this PR further.

**Comment by ZeWaka** [comment]
every day I wake up and dread porting this

File: tgui/packages/common/react.ts (hunk 2 above)
Comment (Mothblocks) [reply tgui/packages/common/react.ts:76]: 👍

**Comment by MrDoomBringer** [comment]
What a shot!

**Comment by ZeWaka** [comment]
fuck

**Comment by Mothblocks** [comment]
at least you wont have to wake up another day

**Comment by SplinterGP** [comment]
pain porting this

**Comment by ZeWaka** [comment]
thanks for showing me how
